File: bnbagent/x402/payer.py
"""X402Payer — the delegated-payment seam (design §3.2, P5).

Where :class:`~bnbagent.x402.signer.X402Signer` exposes the *signing
primitive* ("sign these bytes"), ``X402Payer`` is the seam one level up:
"handle this payment" — the abstraction proven by a decade of L402/aperture
(out-of-process lnd payments behind a maxCost guard) and by the official
x402 ``on_payment_required`` escape hatch. Wallets whose payment machinery
lives outside the SDK (an external CLI, a custodial API) plug in here as a
whole, instead of pretending to be byte signers.

The contract is distilled from two real implementations (the
second-implementation discipline): studio's ``fetch_with_payment`` buyer and
the ``twak x402 quote/request`` CLI client. Implementations may accept extra
keyword arguments beyond the Protocol's signatures (e.g. route-preference
hints); callers relying only on the Protocol stay portable.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Protocol, runtime_checkable


@dataclass(frozen=True)
class X402PaymentOption:
    """One payable route from a 402 challenge (a quote ``accepts`` entry).

    Field-verified against ``twak x402 quote --json`` output (design §10.2):
    the CLI pre-filters routes its client cannot pay, so every option here
    is nominally payable by the backing wallet.
    """

    #: CAIP-2 network identifier, e.g. ``"eip155:56"``.
    network: str
    #: x402 scheme, e.g. ``"exact"``.
    scheme: str
    #: Token contract address. For EIP-3009 routes this address is also the
    #: EIP-712 domain ``verifyingContract``.
    asset: str
    token_name: str | None
    #: Price in atomic token units (parsed from the CLI's decimal string).
    amount: int
    pay_to: str
    #: e.g. ``"eip3009"`` or ``"permit2"``.
    transfer_method: str | None
    #: The challenge's claimed payment-validity window, in seconds.
    max_timeout_seconds: int | None
    preferred: bool
    requires_approval: bool
    description: str | None

    @classmethod
    def from_cli(cls, entry: dict) -> X402PaymentOption:
        """Map a camelCase CLI ``accepts`` entry; missing optionals → None/False."""
        timeout = entry.get("maxTimeoutSeconds")
        return cls(
            network=str(entry["network"]),
            scheme=str(entry.get("scheme", "exact")),
            asset=str(entry["asset"]),
            token_name=entry.get("tokenName"),
            amount=int(entry["amount"]),
            pay_to=str(entry["payTo"]),
            transfer_method=entry.get("transferMethod"),
            max_timeout_seconds=int(timeout) if timeout is not None else None,
            preferred=bool(entry.get("preferred", False)),
            requires_approval=bool(entry.get("requiresApproval", False)),
            description=entry.get("description"),
        )


@dataclass(frozen=True)
class X402Quote:
    """A parsed 402 challenge: the resource plus its payable routes.

    ``accepts`` may be empty — the quoting client filters out routes on
    chains it cannot pay (field-verified: x402.org on base-sepolia yields
    ``accepts: []`` from the twak client).
    """

    url: str
    description: str | None
    mime_type: str | None
    accepts: tuple[X402PaymentOption, ...]
    summary: str | None
    #: The raw parsed CLI/HTTP quote JSON, for fields not modeled here.
    raw: dict

    @classmethod
    def from_cli(cls, data: dict) -> X402Quote:
        resource = data.get("resource") or {}
        return cls(
            url=str(resource.get("url", "")),
            description=resource.get("description"),
            mime_type=resource.get("mimeType"),
            accepts=tuple(
                X402PaymentOption.from_cli(entry)
                for entry in data.get("accepts") or ()
            ),
            summary=data.get("summary"),
            raw=data,
        )


@dataclass(frozen=True)
class X402PaymentResult:
    """Outcome of a delegated x402 payment.

    ``response`` is the paid endpoint's response body **verbatim** —
    arbitrary JSON, exactly what an unpaid request would have returned.

    The payment metadata fields are Optional **by design**: the twak CLI
    does not surface a settlement receipt in stdout (gaps S-7 — the
    ``PAYMENT-RESPONSE`` header is consumed internally and only echoed
    human-readably on stderr). Delegated payers therefore fill
    ``amount``/``asset``/``network``/``pay_to`` from the **quoted** option
    they paid against, not from settlement. ``transaction`` is best-effort:
    some endpoints echo a tx hash in their own body (e.g. pieverse's
    ``tx_hash``), most don't.
    """

    success: bool
    #: The endpoint's response body, verbatim.
    response: Any
    amount: int | None = None
    asset: str | None = None
    network: str | None = None
    pay_to: str | None = None
    transaction: str | None = None


@runtime_checkable
class X402Payer(Protocol):
    """Structural contract for delegated x402 payment backends.

    Two methods, aligned with both the CLI verbs and x402 semantics —
    deliberately not ``pay()``: a cache hit on ``request`` may not pay at
    all. Implementations may accept extra keyword arguments.
    """

    def quote(
        self, url: str, *, method: str = "GET", body: str | None = None
    ) -> X402Quote:
        """Fetch the 402 challenge for ``url`` without paying."""
        ...

    def request(
        self,
        url: str,
        *,
        max_payment: int,
        method: str = "GET",
        body: str | None = None,
    ) -> X402PaymentResult:
        """Fetch ``url``, completing an x402 payment up to ``max_payment``
        atomic units if challenged."""
        ...

File: typescript/src/x402/payer.ts
/**
 * X402Payer — the delegated-payment seam.
 *
 * Where `X402Signer` exposes the *signing primitive* ("sign these bytes"),
 * `X402Payer` is the seam one level up: "handle this payment". Wallets
 * whose payment machinery lives outside the SDK (an external CLI, a
 * custodial API) plug in here as a whole, instead of pretending to be byte
 * signers.
 *
 * Port of `python/bnbagent/x402/payer.py`. TWAK integration itself
 * (`python/bnbagent/x402/twak.py`) is out of scope for this port — only the
 * data types and the CLI-shaped parse helpers are ported here.
 */

import { resolveB402Asset } from "./assets.js";
import type { ExpectedB402Asset } from "./assets.js";

/**
 * One payable route from a 402 challenge (a quote `accepts` entry).
 *
 * Field-verified against `twak x402 quote --json` output: the CLI
 * pre-filters routes its client cannot pay, so every option here is
 * nominally payable by the backing wallet.
 */
export interface X402PaymentOption {
  /** CAIP-2 network identifier, e.g. `"eip155:56"`. */
  network: string;
  /** x402 scheme, e.g. `"exact"`. */
  scheme: string;
  /**
   * Token contract address. For EIP-3009 routes this address is also the
   * EIP-712 domain `verifyingContract`.
   */
  asset: string;
  tokenName?: string;
  /** Price in atomic token units (parsed from the CLI's decimal string). */
  amount: bigint;
  payTo: string;
  /** e.g. `"eip3009"` or `"permit2"`. */
  transferMethod?: string;
  /** The challenge's claimed payment-validity window, in seconds. */
  maxTimeoutSeconds: number | null;
  preferred: boolean;
  requiresApproval: boolean;
  description?: string;
}

/** Throws if `entry[field]` is missing/null/undefined; mirrors Python's KeyError. */
function requireField(entry: Record<string, unknown>, field: string): unknown {
  const value = entry[field];
  if (value === undefined || value === null) {
    throw new Error(`x402 payment option missing required field: ${field}`);
  }
  return value;
}

/**
 * Map a camelCase CLI `accepts` entry; missing optionals → undefined/false.
 *
 * `network`, `asset`, `amount`, and `payTo` are required (mirrors Python's
 * `entry["..."]` KeyError behavior) — a missing value throws immediately
 * rather than silently coercing to the string `"undefined"`.
 */
export function paymentOptionFromCli(
  entry: Record<string, unknown>,
): X402PaymentOption {
  const timeout = entry.maxTimeoutSeconds;
  return {
    network: String(requireField(entry, "network")),
    scheme: String(entry.scheme ?? "exact"),
    asset: String(requireField(entry, "asset")),
    tokenName: entry.tokenName as string | undefined,
    amount: BigInt(requireField(entry, "amount") as bigint | number | string),
    payTo: String(requireField(entry, "payTo")),
    transferMethod: entry.transferMethod as string | undefined,
    maxTimeoutSeconds:
      timeout !== undefined && timeout !== null ? Number(timeout) : null,
    preferred: Boolean(entry.preferred ?? false),
    requiresApproval: Boolean(entry.requiresApproval ?? false),
    description: entry.description as string | undefined,
  };
}

/** Resolve the option's exact `network + asset` through the catalog. */
export function expectedAssetFromPaymentOption(
  option: X402PaymentOption,
): ExpectedB402Asset {
  return resolveB402Asset(option.network, option.asset);
}

/**
 * A parsed 402 challenge: the resource plus its payable routes.
 *
 * `accepts` may be empty — the quoting client filters out routes on chains
 * it cannot pay.
 */
export interface X402Quote {
  url: string;
  description?: string;
  mimeType?: string;
  accepts: readonly X402PaymentOption[];
  summary?: string;
  /** The raw parsed CLI/HTTP quote JSON, for fields not modeled here. */
  raw: Record<string, unknown>;
}

export function quoteFromCli(data: Record<string, unknown>): X402Quote {
  const resource = (data.resource as Record<string, unknown> | undefined) ?? {};
  const accepts = (data.accepts as Record<string, unknown>[] | undefined) ?? [];
  return {
    url: String(resource.url ?? ""),
    description: resource.description as string | undefined,
    mimeType: resource.mimeType as string | undefined,
    accepts: accepts.map(paymentOptionFromCli),
    summary: data.summary as string | undefined,
    raw: data,
  };
}

/**
 * Outcome of a delegated x402 payment.
 *
 * `response` is the paid endpoint's response body **verbatim**. The
 * payment metadata fields are optional by design — delegated payers fill
 * `amount`/`asset`/`network`/`payTo` from the **quoted** option they paid
 * against, not from settlement. `transaction` is best-effort.
 */
export interface X402PaymentResult {
  success: boolean;
  /** The endpoint's response body, verbatim. */
  response: unknown;
  amount?: bigint;
  asset?: string;
  network?: string;
  payTo?: string;
  transaction?: string;
}

export type X402TransferMethod = "eip3009" | "permit2-exact";

interface ExpectedX402RouteBase {
  readonly x402Version: 2;
  readonly scheme: "exact";
  readonly network: `eip155:${number}`;
  readonly asset: string;
  readonly amount: bigint;
  readonly payTo: string;
  readonly maxTimeoutSeconds: number;
}

export type ExpectedX402Route =
  | (ExpectedX402RouteBase & {
      readonly transferMethod: "eip3009";
    })
  | (ExpectedX402RouteBase & {
      readonly transferMethod: "permit2-exact";
      readonly name: string;
      readonly version: string;
      /** Expected B402 proxy from the caller's trusted capability snapshot. */
      readonly spenderAddress: string;
      /** Explicit trust root; the challenge cannot add to this list. */
      readonly trustedSpenders: readonly string[];
    });

export interface X402ExactPaymentResult extends X402PaymentResult {
  readonly success: true;
  readonly amount: bigint;
  readonly asset: string;
  readonly network: `eip155:${number}`;
  readonly payTo: string;
  readonly transferMethod: X402TransferMethod;
  readonly spenderAddress?: string;
}

export interface X402ExactRequestOptions {
  readonly expectedRoute: ExpectedX402Route;
  readonly maxPayment: bigint;
  readonly method?: string;
  readonly body?: string;
}

/**
 * Structural contract for delegated x402 payment backends.
 *
 * Two methods, aligned with both the CLI verbs and x402 semantics —
 * deliberately not `pay()`: a cache hit on `request` may not pay at all.
 * Implementations may accept extra options.
 */
export interface X402Payer {
  /** Methods this implementation can bind atomically in `requestExact`. */
  readonly exactTransferMethods?: readonly X402TransferMethod[];

  /** Fetch the 402 challenge for `url` without paying. */
  quote(
    url: string,
    opts?: { method?: string; body?: string },
  ): Promise<X402Quote>;

  /**
   * Fetch `url`, completing an x402 payment up to `maxPayment` atomic units
   * if challenged.
   */
  request(
    url: string,
    opts: { maxPayment: bigint; method?: string; body?: string },
  ): Promise<X402PaymentResult>;

  /**
   * Fetch, validate, sign, and retry one challenge without re-fetching or
   * selecting a route outside the caller-supplied exact binding.
   */
  requestExact?(
    url: string,
    opts: X402ExactRequestOptions,
  ): Promise<X402ExactPaymentResult>;
}
